Make sure built-in procedures close resources when done

systay opened the pull request:
The first two commits are backports from 3.3 and should be forward-merged with a null commit. The last commit is applicable to both 3.2 and 3.3

henriknyman committed:
Fix statement acquisition in token accessing built-in procedures

The built-in procedures using `TokenAccess` (`db.labels()`,
`db.relationshipTypes()`, `db.propertyKeys()`) would close their
acquired `KernelStatement` prematurely before returning an iterator
stream was using the statement and that would close it when exhausted.
This caused a problem when a multiple of these procedure calls were
made in the same query. Changed this so that ownership of the acquired
statement reference is transferred to the iterator (which is consistent
with the core API call usage of `TokenAccess`).

Also changed the `TokenIterator.close()` to be idempotent, since
exhausting the procedure stream also causes close to be invoked.

henriknyman committed:
Fix closing of statement on profiler not ready exception

When trying to get the profile from a Cypher query result before the
result is materialized a `ProfilerStatisticsNotReadyException` is
thrown. This commit makes sure that the result is properly closed
first, so that we do not leak kernel statements.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java:
```diff
@@ -67,17 +67,20 @@
         {
-            return TokenAccess.PROPERTY_KEYS.inUse( statement ).map( PropertyKeyResult::new ).stream();
+            // Ownership of the reference to the acquired statement is transfered to the returned iterator stream
+            return asList( TokenAccess.PROPERTY_KEYS.inUse( statement ).map( PropertyKeyResult::new ) ).stream();
```
Yup, this eagerness here is the preferred way to go

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/TokenAccess.java:
```diff
@@ -143,7 +143,11 @@ private static boolean hasAny( Iterator<?> iter )
         @Override
         public void close()
         {
-            statement.close();
```
Is this strictly necessary or just being over the top safe?

systay replied:
This was back-ported from the 3.3 branch. Not sure - what do you think, @henriknyman?

henriknyman replied:
Hmm I don't remember.. it could be the latter.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureCompiler.java:
```diff
@@ -671,7 +671,11 @@ public boolean hasNext() throws ProcedureException
             public void close() throws Exception
             {
-                closeable.close();
+                if ( closeable != null )
```
Is this strictly necessary or just being over the top safe?

thobe commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/ProcedureCallSupportAcceptanceTest.scala:
```diff
@@ -87,4 +87,26 @@ class ProcedureCallSupportAcceptanceTest extends ProcedureCallAcceptanceTest {
     ))
   }
+
+  test("should be able to execute union of multiple token accessing procedures") {
```
This test fails with:

```
org.neo4j.graphdb.NotInTransactionException: The statement has been closed.
```

thobe commented:
Just to verify my understanding of this changeset. What is supposed to fix the issue is mainly:

- [Line 102 in DefaultExecutionResultBuilderFactory](https://github.com/neo4j/neo4j/pull/10071/files#diff-9e953639cb3fbc71d1fa8d742723cce0R102)
- [Line 102 in ProcedureExecutionResult](https://github.com/neo4j/neo4j/pull/10071/files#diff-1257d865b16a46ad191c188d40ca02ccR102), and
- [Line 54 in CompiledExecutionResult](https://github.com/neo4j/neo4j/pull/10071/files#diff-123545a8c821d04dbcd74857dc4c0545R54)

Then there are changes to some of the procedures to ensure that they finish their reads before the statement is closed.

Is that correct?

systay commented:
We are trying to make the test `should be able to execute union of multiple token accessing procedures` go green, and those are the changes needed.

henriknyman commented:
@thobe The three things you mention are a secondary issue when we run in PROFILE mode. The main issue is that the three built-in procedures could not close the statement while returning a lazy stream.

henriknyman commented:
@tinwelint The reason we have to transfer ownership of the statement into `TokenAccess` is because that is the way the Core API works. The statement will be closed when the iterator/stream is exhausted.
We could make the built-in procedures eager if we are worried about leaking a statement in case of an exception, but we cannot do try-with-resource since the statement will be closed by exhausting the stream in the happy case.

Commit:
Make sure to eagerly consume iterators from kernel

This way, we can catch any exceptions and close the statement instead
of leaving it open one failures.

tinwelint commented:
@henriknyman OK right, so that will only get closed in the happy case (no exception and stream fully consumed) if letting the stream manage the closing of the resource. This eagerness feels better and is typically on very small number of tokens.

thobe commented:
@systay / @henriknyman: this is still not where it needs to be:

```
java.lang.AssertionError: Failed on procedure CALL db.propertyKeys()
    at org.neo4j.kernel.builtinprocs.ProcedureResourcesIT.verifyProcedureCloseAllAcquiredKernelStatements(ProcedureResourcesIT.java:118)
    at org.neo4j.kernel.builtinprocs.ProcedureResourcesIT.allProcedures(ProcedureResourcesIT.java:82)
```

henriknyman committed:
Close kernel statements on the eager built-in procedures

Use try-with-resource on kernel statement in the token accessing
built-in procedures.
